Modules/_io/_iomodule.c

```diff
@@ -479,7 +480,13 @@ _io_text_encoding_impl(PyObject *module, PyObject *encoding, int stacklevel)
                 return NULL;
             }
         }
-        return &_Py_ID(locale);
+        const PyPreConfig *preconfig = &_PyRuntime.preconfig;
+        if (preconfig->utf8_mode) {
+            encoding = &_Py_STR(utf_8);
```

Comment:
`_Py_DECLARE_STR()` should always precede use of `_Py_STR()`, even if the same string is declared elsewhere (e.g. `anon_string` in Python/pythonrun.c).  That way things don't break when `_Py_DECLARE_STR()` is dropped elsewhere.  It also makes it clear what the actual string is at the place it's used.  `_Py_DECLARE_STR()` is an empty macro so there is no problem from having duplicates, and Tools/scripts/generate_global_objects.py prevents mismatches.  Sorry all that wasn't more clear.

```suggestion
            _Py_DECLARE_STR(utf_8, "utf-8");
            encoding = &_Py_STR(utf_8);
```

Reply:
FYI, I added a comment for _Py_DECLARE_STR() in gh-32063.